code/graphics/opengl/gropengltexture.cpp

```diff
@@ -1554,8 +1554,13 @@ static void opengl_free_fbo_slot(int id) {
 
 	Assertion(fbo != nullptr, "Invalid id passed to opengl_free_fbo_slot!");
 
+	//We need to reset the FBO but keep it's ID, otherwise they'll get reassigned with ID 0
+	int fbo_id = fbo->fbo_id;
+
 	// Reset this slot using the default constructor
 	*fbo = fbo_t();
+
+	fbo->fbo_id = fbo_id;
```

Comment:
This is still a memory leak since that FBO slot will never be reused if the id is kept as a non-zero value.

Reply:
Ok, that fixed the memory leak but can you go into detail a bit why this is necessary? It's not good to basically leave dangling references around.

Reply:
So, the issue is that the code makes two assumptions:
A released FBO Object is immediately good to go (hence the return &target; once it finds a free buffer), and that new ID's that are assigned are always increased by one.
There are three options to solve this:
1. Don't save old, freed FBO pointer, and never assign another FBO object with the ID this object once had (which now would be free)
2. Keep the old, freed FBO pointer, create a fresh fbo object in it (that was what the code did before), and keep the ID this pointer has (that's what I changed, otherwise all kept fbo's would have ID 0, which is bad).
3. Keep the old, freed FBO pointer, create a fresh fbo object in it, and once it is reused, assign it a free ID. I didn't do this, as it'd either mean iterating through FBO Objects until we find an ID that is free (and given this is techincally an unordered vector, that's expensive), or keeping track of freed ID's.

Reply:
TL;DR, assigning the new object this ID is less of a "dangling reference" and more of a: We reset the FBO Object that has this ID.
The proper way would be to save FBO objects as a map with their ID as the key, but that'd be a major overhaul of the whole system

Reply:
You described what is going on but now _why_ your changes are the correct ones. What does keeping the id change in the runtime behavior that fixes rendering of detached render targets?

Reply:
Ah, my bad.
The next rendertarget being created after an detachment get's assigned the old FBO Object.
Without the change, it'll always have the ID 0 (which is assumed to then be a non-existing FBO, see check that prints "Tried to get an OpenGL FBO that didn't exist!" when ID = 0). And that means, that setTarget(texture) will fail, as the FBO doesn't have a legit ID
With the change, the FBO Object has a legit and unique ID, so that setting that texture as a target actually works.
So it's less that the texture can't be rendered without this change, but rather that it can't be rendered _to_, and is thus always black or transparent (either of which AABitmap then displays as transparent, making it seem like it had never rendered)

Reply:
Ahh, now I see. Thanks for the explanation and the detective work here!
However, with that in mind then it would be more correct to change `opengl_get_free_fbo` so that a new FBO ID is created (using `next_fbo_id`) instead of leaving it at `-1` which is just completely wrong.

I guess so far noone bothered to clean up their render targets so this bug was never discovered...

Reply:
If you always assign a new ID, then you will end up with ID's of freed up FBO's being unused, rather than reused like this. Would that be preferable to my current solution?

Reply:
At least in my opinion, that is the preferred solution since it simplifies the cleanup code and keeps the FBO structs in a "default" state when they are not in use.